app/src/main/java/com/example/campusguide/search/PointsOfInterest.kt

```diff
@@ -0,0 +1,130 @@
+import com.example.campusguide.R
+import com.example.campusguide.directions.DirectionsFlow
+import com.example.campusguide.location.Location
+import com.example.campusguide.location.LocationProvider
+import com.example.campusguide.map.GoogleMapAdapter
+import com.example.campusguide.search.outdoor.PlacesApiSearchResultProvider
+import com.google.android.libraries.places.api.Places
+import com.google.android.libraries.places.api.model.Place
+import com.google.android.libraries.places.api.net.FetchPlaceRequest
+import com.google.android.libraries.places.api.net.FetchPlaceResponse
+import com.google.android.libraries.places.api.net.PlacesClient
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.launch
+
+class PointsOfInterest(
+    private val activity: MapsActivity,
+    private val locationProvider: LocationProvider,
+    private val map: GoogleMapAdapter,
+    private val directionsFlow: DirectionsFlow
+) : DialogFragment() {
+
+    private val placesClient: PlacesClient
+    var locationListener: SearchLocationListener? = null
+
+    private val nearbyLocations = PlacesApiSearchResultProvider(activity)
+
+    init {
+        if (!Places.isInitialized())
+            Places.initialize(
+                activity.applicationContext,
+                activity.getString(R.string.google_maps_key)
+            )
+
+        placesClient = Places.createClient(activity)
+    }
+
+    override fun onCreateView(
+        inflater: LayoutInflater,
+        container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View? {
+        dialog?.window?.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
+        val view = inflater.inflate(R.layout.points_of_interest_options, container, false)
+
+        view.findViewById<Button>(R.id.cafe)?.setOnClickListener {
+            dismiss();
+            searchPOINearCurrentLocation("Cafe");
+        }
+        view.findViewById<Button>(R.id.restaurant)?.setOnClickListener {
+            dismiss();
+            searchPOINearCurrentLocation("Restaurant");
+        }
+        view.findViewById<Button>(R.id.shopping)?.setOnClickListener {
+            dismiss();
+            searchPOINearCurrentLocation("Shopping mall");
+        }
+        view.findViewById<Button>(R.id.pharmacy)?.setOnClickListener {
+            dismiss();
+            searchPOINearCurrentLocation("Pharmacy");
+        }
+        view.findViewById<Button>(R.id.grocerystore)?.setOnClickListener {
+            dismiss();
+            searchPOINearCurrentLocation("Grocery store");
+        }
+        return view
+    }
+
+    /**
+     * Gets the user's last known most recent location
+     */
+
+    private fun searchPOINearCurrentLocation(poi: String) {
+        return locationProvider.getLocation {
+            GlobalScope.launch {
+                getNearbyPlaces(poi, it)
+            }
+        }
+    }
+
+    private suspend fun getNearbyPlaces(poi: String, location: Location) {
```

Comment:
Why is this a suspend function?

Reply:
this was really complicated but it needs to be a suspend function or else I can't bring the places response out to the main Point of Interest class. 
I peer programmed this with Felix and he helped me figure it out